### PR TITLE
Introduce Pulsar Dashboard

### PR DESCRIPTION
**Pull request by merlimat**
### Motivation

Add a web dashboard to report the usages across all topics / namespaces / brokers.

### Modifications

Introduce metrics collector, Django web app and Docker configuration for the Pulsar dashboard.

### CODE REVIEW

**Comment by agarman** [comment]
We had setup Prometheus. We'll be happy to switch.

**Comment by merlimat** [comment]
@agarman This is more like solution to display tabular sortable data for each topic. The is no historic metric right now, just the last interval values. 

This is kind of complementary to having time series graphs, perhaps at the namespace level granularity.

**Comment by agarman** [comment]
We weren't using historic data from Prometheus...the metrics were driving grafana dashboards. I'll check to see if we can share the Prometheus + grafana setup via docker.

**Comment by merlimat** [comment]
That would be great to have an out of the box default dashboard and collection mechanism.

File: docs/Dashboard.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+## Install
+
+The easiest way to use the dashboard is to run it inside
+a Docker container. A `Dockerfile` to generate the image
+is provided.
+
+To generate the Docker image:
+
+```shell
+$ docker build -t pulsar-dashboard dashboard
```
Comment (saandrews) [inline docs/Dashboard.md]: Add apt-get install docker before this step?

File: docs/Dashboard.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+To run the dashboard:
+
+```shell
+$ MY_SERVICE_URL=http://broker.example.com:8080/
+$ docker run --rm -it -p 80:80 \
+    -e SERVICE_URL=$MY_SERVICE_URL\
+     pulsar-dashboard
+```
+
```
Comment (saandrews) [inline docs/Dashboard.md]: Should we call out the endpoint to get the stats? e.g: http://host:80/admin/clusters

File: docs/Dashboard.md (hunk 1 above)
Comment (merlimat) [inline docs/Dashboard.md]: You can actually do that on mac/ubuntu/rhel.. so I wouldn't use necessarily `apt-get` in the instructions. I think at this point everyone knows about Docker, though we can put a link to the homepage.

File: docs/Dashboard.md (hunk 2 above)
Comment (merlimat) [inline docs/Dashboard.md]: You mean, where to point the browser to access the dashboard? Or from where the collector pulls that stats from?

File: docs/Dashboard.md (hunk 1 above)
Comment (saandrews) [inline docs/Dashboard.md]: Ok, let us get the link to docker documentation(https://docs.docker.com/engine/installation/). Assuming everyone knows about docker is debatable :)

File: docs/Dashboard.md (hunk 2 above)
Comment (saandrews) [inline docs/Dashboard.md]: Yes, the end point in the browser to access dashboard.

File: docs/Dashboard.md (hunk 1 above)
Comment (merlimat) [inline docs/Dashboard.md]: @saandrews Updated documentation

File: dashboard/django/stats/templates/stats/clusters.html [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+            </thead>
+            <tbody>
+            {% for consumerStats in consumers %}
+                <tr>
+                    <td>{{consumerStats.consumerName}}</td>
+                    <td>{{consumerStats.address}}</td>
+                    <td>{{consumerStats.msgRateOut | intcomma}}</td>
+                    <td>{{consumerStats.msgThroughputOut | intcomma}}</td>
+                    <td>{{consumerStats.msgRateRedeliver | intcomma}}</td>
+                    <td title="{{consumerStats.connectedSince}}">{{consumerStats.connectedSince | naturaltime}}</td>
```
Comment (rdhabalia) [inline dashboard/django/stats/templates/stats/clusters.html]: should we give timestamp instead of `naturaltime` incase if consumer is connected since 20 days then it will be easy for user to understand the exact date rather `20 days ago`.

File: dashboard/django/stats/templates/stats/brokers.html [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+{% block title %}Brokers{% endblock %}
+
+{% block content %}
+
+<div class="module filtered" id="changelist">
+<div id="changelist-filter">
+    <h2>Clusters</h2>
+<ul>
+   <li {% if not selectedCluster %}class="selected"{% endif %}>
```
Comment (rdhabalia) [inline dashboard/django/stats/templates/stats/brokers.html]: here it seems, it doesn't show broker with 0 bundles. broker has to own at least one bundle to appear here. should we show regardless number of bundle.

**Comment by rdhabalia** [comment]
@merlimat Just minor comments. It looks good to me and great effort.

File: dashboard/django/stats/templates/stats/brokers.html [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        {% column_header brokers 'throughputOut' 'Mbps Out' %}
+        {% column_header brokers 'backlog' 'Backlog' %}
+    </tr>
+</thead>
+<tbody>
+
+{% for broker in brokers.results %}
+    <tr class="{% cycle 'row1' 'row2' %}">
+        <th>
+            <a href="{% url 'broker' broker.url %}">{{broker.url}}</a>
```
Comment (merlimat) [inline dashboard/django/stats/templates/stats/brokers.html]: Yep, I think I kind of forgot about the broker view part :)

File: dashboard/django/stats/templates/stats/clusters.html (hunk 3 above)
Comment (merlimat) [inline dashboard/django/stats/templates/stats/clusters.html]: The actual datetime is shown when you hover the natural time

File: dashboard/django/stats/templates/stats/brokers.html (hunk 5 above)
Comment (rdhabalia) [inline dashboard/django/stats/templates/stats/brokers.html]: or should we merge this PR and broker-view can be part of separate PR?

File: dashboard/django/stats/templates/stats/brokers.html (hunk 4 above)
Comment (merlimat) [inline dashboard/django/stats/templates/stats/brokers.html]: Fixed it

**Comment by merlimat** [comment]
@rdhabalia Updated the PR

Added:
 * Replication views
 * Show brokers even with no topic
 * Added single broker view

**Comment by merlimat** [comment]
@saandrews @rdhabalia Ping
